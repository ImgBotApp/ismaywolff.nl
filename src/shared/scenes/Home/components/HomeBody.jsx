import React from 'react'
import { number, object, objectOf, shape, bool, string, arrayOf } from 'prop-types'
import { Thumbnail, ThumbnailGrid } from '../../../components/thumbnail'
import { Title, TitleWithLine } from '../../../components/text'
import { AppError } from '../../../components/errors'
import { Spinner } from '../../../components/spinner'
import { Link } from '../../../components/link'
import { Hero } from '../../../components/hero'
import { Box } from '../../../components/box'

const HomeBody = ({ works, images, workEntities, imageEntities, featured, hero }) => {
  const fetchingWorks = works.isFetching || !works.lastUpdated
  const fetchingImages = images.isFetching || !images.lastUpdated
  const worksError = works.errorMessage
  const imagesError = images.errorMessage

  if (fetchingWorks || fetchingImages) {
    return <Spinner />
  }

  if (worksError || imagesError) {
    return <AppError errorMessage={worksError || imagesError} />
  }

  // For multiple heros an item is picked randomly
  const heroId = hero[Math.floor(Math.random() * hero.length)]

  return (
    <div>
      <Hero work={workEntities[heroId]} image={imageEntities[workEntities[heroId].thumbnail]} />
      <Box margin="var(--size-larger) 0 var(--size-larger) 0">
        <Title tag="h2">Hi there</Title>
        <p>
          I'm an artist, a social scientist and a software engineer. My work is a combination of the
          most inspiring parts of those disciplines and an expression of everything that interests,
          moves or inspires me.
        </p>
        <Box center>
          <Link size="var(--size-small)" to="/about">
            Learn more
          </Link>
        </Box>
      </Box>
      <Box margin="var(--size-larger) 0 0 0">
        <TitleWithLine tag="h2" size="var(--size-small)">
          Featured work
        </TitleWithLine>
        <Box margin="var(--size-regular) 0 var(--size-regular) 0">
          <ThumbnailGrid>
            {featured.map(id =>
              <Thumbnail
                work={workEntities[id]}
                image={imageEntities[workEntities[id].thumbnail]}
                key={id}
              />
            )}
          </ThumbnailGrid>
        </Box>
        <Box center>
          <Link size="var(--size-small)" to="/work">
            See all work
          </Link>
        </Box>
      </Box>
    </div>
  )
}

HomeBody.propTypes = {
  hero: arrayOf(string).isRequired,
  featured: arrayOf(string).isRequired,
  workEntities: objectOf(object).isRequired,
  imageEntities: objectOf(object).isRequired,
  works: shape({
    lastUpdated: number.isRequired,
    errorMessage: string.isRequired,
    isFetching: bool.isRequired,
    result: arrayOf(string).isRequired
  }).isRequired,
  images: shape({
    lastUpdated: number.isRequired,
    errorMessage: string.isRequired,
    isFetching: bool.isRequired,
    result: arrayOf(string).isRequired
  }).isRequired
}

export default HomeBody
