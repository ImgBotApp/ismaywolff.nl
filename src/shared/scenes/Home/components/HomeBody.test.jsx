import React from 'react'
import { shallow } from 'enzyme'
import HomeBody from './HomeBody'

const recentFetch = 1
Date.now = jest.fn(() => recentFetch + 1)

describe('<HomeBody />', () => {
  it('renders correctly', () => {
    const images = {
      lastUpdated: recentFetch,
      errorMessage: '',
      isFetching: false,
      result: ['imageId']
    }
    const works = {
      lastUpdated: recentFetch,
      errorMessage: '',
      isFetching: false,
      result: ['workId']
    }
    const featured = ['workId']
    const hero = ['workId']
    const workEntities = {
      workId: {
        title: 'title',
        slug: 'slug',
        thumbnail: 'imageId',
        type: 'type',
        published: '10-01-2010',
        summary: 'summary',
        images: ['imageId'],
        hero: true,
        featured: true
      }
    }
    const imageEntities = {
      imageId: {
        title: 'title',
        url: 'https://url.com',
        width: 100,
        height: 100
      }
    }

    const wrapper = shallow(
      <HomeBody
        featured={featured}
        hero={hero}
        images={images}
        works={works}
        imageEntities={imageEntities}
        workEntities={workEntities}
      />
    )
    expect(wrapper).toMatchSnapshot()
  })

  it('renders a loading state for works', () => {
    const images = {
      lastUpdated: recentFetch,
      errorMessage: '',
      isFetching: false,
      result: ['imageId']
    }
    const works = {
      lastUpdated: 0,
      errorMessage: '',
      isFetching: true,
      result: []
    }
    const featured = []
    const hero = []
    const workEntities = {}
    const imageEntities = {
      imageId: {
        title: 'title',
        url: 'https://url.com',
        width: 100,
        height: 100
      }
    }

    const wrapper = shallow(
      <HomeBody
        featured={featured}
        hero={hero}
        images={images}
        works={works}
        imageEntities={imageEntities}
        workEntities={workEntities}
      />
    )
    expect(wrapper).toMatchSnapshot()
  })

  it('renders a loading state for images', () => {
    const images = {
      lastUpdated: 0,
      errorMessage: '',
      isFetching: true,
      result: []
    }
    const works = {
      lastUpdated: recentFetch,
      errorMessage: '',
      isFetching: false,
      result: ['workId']
    }
    const featured = ['workId']
    const hero = ['workId']
    const workEntities = {
      workId: {
        title: 'title',
        slug: 'slug',
        thumbnail: 'imageId',
        type: 'type',
        published: '10-01-2010',
        summary: 'summary',
        images: ['imageId'],
        hero: true,
        featured: true
      }
    }
    const imageEntities = {}

    const wrapper = shallow(
      <HomeBody
        featured={featured}
        hero={hero}
        images={images}
        works={works}
        imageEntities={imageEntities}
        workEntities={workEntities}
      />
    )
    expect(wrapper).toMatchSnapshot()
  })

  it('renders work errors', () => {
    const images = {
      lastUpdated: recentFetch,
      errorMessage: '',
      isFetching: false,
      result: ['imageId']
    }
    const works = {
      lastUpdated: recentFetch,
      errorMessage: 'Error while fetching works',
      isFetching: false,
      result: []
    }
    const featured = []
    const hero = []
    const workEntities = {}
    const imageEntities = {
      imageId: {
        title: 'title',
        url: 'https://url.com',
        width: 100,
        height: 100
      }
    }

    const wrapper = shallow(
      <HomeBody
        featured={featured}
        hero={hero}
        images={images}
        works={works}
        imageEntities={imageEntities}
        workEntities={workEntities}
      />
    )
    expect(wrapper).toMatchSnapshot()
  })

  it('renders image errors', () => {
    const images = {
      lastUpdated: recentFetch,
      errorMessage: 'Error while fetching images',
      isFetching: false,
      result: []
    }
    const works = {
      lastUpdated: recentFetch,
      errorMessage: '',
      isFetching: false,
      result: ['workId']
    }
    const featured = ['workId']
    const hero = ['workId']
    const workEntities = {
      workId: {
        title: 'title',
        slug: 'slug',
        thumbnail: 'imageId',
        type: 'type',
        published: '10-01-2010',
        summary: 'summary',
        images: ['imageId'],
        hero: true,
        featured: true
      }
    }
    const imageEntities = {}

    const wrapper = shallow(
      <HomeBody
        featured={featured}
        hero={hero}
        images={images}
        works={works}
        imageEntities={imageEntities}
        workEntities={workEntities}
      />
    )
    expect(wrapper).toMatchSnapshot()
  })
})
