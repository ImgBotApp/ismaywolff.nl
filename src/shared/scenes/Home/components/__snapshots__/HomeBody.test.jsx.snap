// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<HomeBody /> renders a loading state for images 1`] = `<styled.div />`;

exports[`<HomeBody /> renders a loading state for works 1`] = `<styled.div />`;

exports[`<HomeBody /> renders correctly 1`] = `
<div>
  <Hero
    image={
      Object {
        "height": 100,
        "title": "title",
        "url": "https://url.com",
        "width": 100,
      }
    }
    work={
      Object {
        "featured": true,
        "hero": true,
        "images": Array [
          "imageId",
        ],
        "published": "10-01-2010",
        "slug": "slug",
        "summary": "summary",
        "thumbnail": "imageId",
        "title": "title",
        "type": "type",
      }
    }
  />
  <styled.div
    center={false}
    color="unset"
    margin="var(--size-larger) 0 var(--size-larger) 0"
    size="unset"
  >
    <TitleFactory
      center={false}
      margin="unset"
      size="1rem"
      tag="h2"
    >
      Hi there
    </TitleFactory>
    <p>
      I'm an artist, a social scientist and a software engineer. My work is a combination of the most inspiring parts of those disciplines and an expression of everything that interests, moves or inspires me.
    </p>
    <styled.div
      center={true}
      color="unset"
      margin="unset"
      size="unset"
    >
      <Link
        size="var(--size-small)"
        to="/about"
      >
        Learn more
      </Link>
    </styled.div>
  </styled.div>
  <styled.div
    center={false}
    color="unset"
    margin="var(--size-larger) 0 0 0"
    size="unset"
  >
    <TitleWithLineFactory
      center={false}
      margin="unset"
      size="var(--size-small)"
      tag="h2"
    >
      Featured work
    </TitleWithLineFactory>
    <styled.div
      center={false}
      color="unset"
      margin="var(--size-regular) 0 var(--size-regular) 0"
      size="unset"
    >
      <styled.div>
        <Thumbnail
          image={
            Object {
              "height": 100,
              "title": "title",
              "url": "https://url.com",
              "width": 100,
            }
          }
          work={
            Object {
              "featured": true,
              "hero": true,
              "images": Array [
                "imageId",
              ],
              "published": "10-01-2010",
              "slug": "slug",
              "summary": "summary",
              "thumbnail": "imageId",
              "title": "title",
              "type": "type",
            }
          }
        />
      </styled.div>
    </styled.div>
    <styled.div
      center={true}
      color="unset"
      margin="unset"
      size="unset"
    >
      <Link
        size="var(--size-small)"
        to="/work"
      >
        See all work
      </Link>
    </styled.div>
  </styled.div>
</div>
`;

exports[`<HomeBody /> renders image errors 1`] = `
<AppError
  errorMessage="Error while fetching images"
/>
`;

exports[`<HomeBody /> renders work errors 1`] = `
<AppError
  errorMessage="Error while fetching works"
/>
`;
